srcts/src/shiny/bind.ts

```diff
@@ -47,45 +47,57 @@ function valueChangeCallback(
  * immediately invoked function to keep the sets private and not clutter the
  * scope.
  */
-const bindingsRegistery = (() => {
+const bindingsRegistry = (() => {
+  type BindingTypes = "input" | "output";
+
   /**
    * Keyed by binding IDs to the array of each type of binding that ID is associated for in current app state.
    *
    * Ideally the
    * value would be a length 1 array but in some (invalid) cases there could be
    * multiple types for a single ID.
    */
-  type IdToBindingTypes = Map<string, Array<"input" | "output">>;
+  type IdToBindingTypes = Map<string, BindingTypes[]>;
 
   // Main store of bindings.
   const bindings: IdToBindingTypes = new Map();
 
   /**
-   * Checks if the bindings registery is valid. Currently this just checks for
-   * duplicate IDs but in the future could be expanded to check more conditions
-   * @returns ShinyClientError if current ID bindings are invalid, otherwise null
+   * Checks if the bindings registry is valid. Currently this just checks if IDs
+   * are duplicated within a binding typ but in the future could be expanded to
+   * check more conditions.
+   *
+   * @description IDs are allowed to be duplicated across binding types, but
+   * when duplicated within a binding type we report all uses of the ID.
+   * Currently the IDs are typically stored in the bound element's `id`
+   * attribute, in which case they really *should* be globally unique for
+   * accessibility and other reasons. However, in practice our bindings still
+   * work as long as inputs the IDs within a binding type don't overlap.
+   *
+   * @returns ShinyClientError if current ID bindings are invalid, otherwise
+   * returns an ok status.
    */
   function checkValidity():
     | { status: "error"; error: ShinyClientError }
     | { status: "ok" } {
-    const duplicateIds: IdToBindingTypes = new Map();
+    type BindingCounts = { [T in BindingTypes]: number };
+    const duplicateIds = new Map<string, BindingCounts>();
+
+    // count duplicate IDs of each binding type
+    bindings.forEach((idTypes, id) => {
+      const counts: { [T in BindingTypes]: number } = { input: 0, output: 0 };
```

Comment:
Oh we should have removed the `BindingCounts` type since the definition above is more concise and would expand automatically if we added new binding types. Sorry, this change was made via the suggestion interface, so this was an oversight.